### PR TITLE
[review] bitbucket support

### LINKED ISSUE #136: Bitbucket Support
Does renovate support Bitbucket at all? If not is there any plan to add support?

### PR DESCRIPTION
**Pull request by iamstarkov**
if completed will close #136 

TODO:
- [x] create empty functions for bitbucket API so as to pass the lint test
- [x] document which functions aren't necessary for v1 of BB support
- [x] implement functions
- [x] create issues for all remaining functions
- [x] remove temporary test script
- [x] implement `utils.recursive` 
- [x] implement MVP-non-essential `// TODO` methods
  - [x] setBaseBranch
  - [x] getAllRenovateBranches
  - [x] setBranchStatus
  - [x] deleteBranch
  - [x] mergeBranch
  - [x] ensureIssue
  - [x] ensureIssueClosing
  - [x] addAssignees
  - [x] addReviewers
  - [x] ensureComment
  - [x] ensureCommentRemoval
  - [x] getCommitMessages

### CODE REVIEW

**Comment by rarkins** [comment]
Great, I will take a look and plan next steps soon

**Comment by iamstarkov** [comment]
on my side, i will add more endpoints

**Comment by iamstarkov** [comment]
@rarkins i sent an collaboration invite to my fork, so can tweak it

**Comment by iamstarkov** [comment]
I added initRepo, non-recursive getFileList (TODO) and getPrList

**Comment by iamstarkov** [comment]
will continue tomorrow

File: lib/platform/bitbucket/index.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+// Search
+
+// Get full file list
+async function getFileList(repoFullName, branchName) {
+  // bitbucket doesnt support recursion for this endpoint
+  // https://developer.atlassian.com/bitbucket/api/2/reference/resource/repositories/%7Busername%7D/%7Brepo_slug%7D/src/%7Bnode%7D/%7Bpath%7D
+  const res = await bbGot(`/repositories/${repoFullName}/src`);
+  // const values = res.body.values;
+
+  // TODO: add support for recursive listing
```
Comment (rarkins) [inline lib/platform/bitbucket/index.js]: That's a real pain that BitBucket doesn't support any recursive file listing. It's pretty essential that Renovate can get a full list of:
- files
- Pull Requests
- Branches

Comment (rarkins) [reply lib/platform/bitbucket/index.js]: @iamstarkov check out https://bitbucket.org/!api/1.0/repositories/iamstarkov/phenomic/directory/master :) 

Found it mentioned here: https://bitbucket.org/site/master/issues/8316/rest-api-list-all-files-in-repo-like-git

You'll need to adjust your bb wrapper somehow to allow this URL to be queried but I think it's worth it.

**Comment by rarkins** [comment]
@iamstarkov I've added empty functions to `lib/platform/bitbucket/index.js`

Ones that throw an error need implementing for BitBucket support to get off the ground. Ones with `// TODO` are ones that I think can be left until later, i.e. not essential functionality for an initial release.

File: lib/platform/bitbucket/index.js (hunk 1 above)
Comment (iamstarkov) [reply lib/platform/bitbucket/index.js]: isnt it the 1st version of the api? i though it will be deprecated

Comment (iamstarkov) [reply lib/platform/bitbucket/index.js]: @rarkins  do we need full file system tree of the repo for the monorepos?

Comment (rarkins) [reply lib/platform/bitbucket/index.js]: Full file system would be necessary for monorepos only. So in theory you could skip it in initial release if you list monorepos as unsupported. However I'd prefer to try/catch the API v1 and fall back to API v2 with a warning, as that way you gain something with essentially zero risk of the API gets revoked before they add recursive support to latest API

Comment (iamstarkov) [reply lib/platform/bitbucket/index.js]: i think we can reduce amount of api calls, by detecting monorepos and executing recursive scheme only for those repos

Comment (iamstarkov) [reply lib/platform/bitbucket/index.js]: and i think its applicable for other platforms as well. maybe not in terms of API hits, but in terms of network bandwidth and latency/speed

**Comment by rarkins** [comment]
@iamstarkov I got notifications of some comments last night but they now seem to be "hidden" by GitHub after code changes.

I don't think you can reliably "detect" monorepos. There are some cases like lerna or yarn workspaces where you might find out its a monorepo based on riles in the root directory. However even in those cases it is allowed that someone starts a yarn workspace in a nested location, e.g. "backend/package.json". And when we say "monorepo", in practice it just means "a repo with package files located outside the root directory", such as someone who puts their `Dockerfile` inside "docker/", for example.

In terms of performance/api calls and other platforms, the others are pretty efficient. GitHub returns the full (recursed) file list in a single API call, while GitLab returns a recused result paginated 100 at a time, which is also not bad. I certainly wouldn't want to "optimise" that if it meant losing the ability to reliably detect package files in subdirectories.

**Comment by rarkins** [comment]
@iamstarkov are there any blockers I can help with to move things forward?

**Comment by iamstarkov** [comment]
@rarkins are `ensureIssue` and `ensureIssueClosing` essential for MVP?

**Comment by iamstarkov** [comment]
@rarkins MVP support has been added

**Comment by rarkins** [comment]
@iamstarkov can I test it against BitBucket cloud using a personal account? Any chance you can add directions for that to this PR description and/or a link into a source file if described there?

**Comment by iamstarkov** [comment]
TODOs methods and recursive helper left

**Comment by iamstarkov** [comment]
@rarkins  im not sure authorization token does work for BB. for personal use change that header to basic auth token in a `./bb-got` file, hints are in https://gist.github.com/iamstarkov/725d20bc5fbf402b785eba06ff9c04c1. and follow instructions how to get token in the README

**Comment by iamstarkov** [comment]
about token header i thought to verify that it works when we will have renovate's BB cloud app

**Comment by rarkins** [comment]
Ideally we can run it from cli, eg like `yarn start —token=abc —platform=bitbucket —autodiscover`. How far off would that be?

**Comment by iamstarkov** [comment]
ii is not far off. does current token implementation support it? if its not we change the implementation in `bb-got` file

**Comment by rarkins** [comment]
> does current token implementation support it?

Do you mean your token implementation or my token implementation prior to bitbucket?

**Comment by iamstarkov** [comment]
i mean, there are different types of tokens: basic, oauth and jwt

**Comment by iamstarkov** [comment]
i dont understand your question, but i referred to token handling in the `bb-got`

**Comment by iamstarkov** [comment]
also we have TODO:

// TODO: get in touch with @rarkins about lifting up the caching into the app layer

File: lib/config/definitions.js [hunk 2]
```diff
@@ -159,6 +159,20 @@ const options = [
     type: 'string',
     admin: true,
   },
+  {
+    name: 'user',
```
Comment (rarkins) [inline lib/config/definitions.js]: I think let's go with verbose "username" and "password" options

File: lib/platform/bitbucket/README.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
+# BitBucket platform
+
+## Credentials setup
+
+### How to get
+
+BB_TOKEN for now is base64 encoded string of your `username:bbaAppPassword`.
```
Comment (rarkins) [inline lib/platform/bitbucket/README.md]: This section can be updated now

**Comment by rarkins** [comment]
My todo's from testing:

- [ ] it "updates" the onboarding PR unnecessarily every time it runs (probably a problem with comparing PR bodies)

- [ ] error getting git tree of onboarding branch

```
DEBUG: GET /2.0/repositories/rhysarkins/test1/src/renovate/configure/?pagelen=100 (repository=rhysarkins/test1)
 INFO: Error retrieving git tree - no files detected (repository=rhysarkins/test1)
```

File: lib/config/definitions.js (hunk 2 above)
Comment (iamstarkov) [reply lib/config/definitions.js]: can i have short variants of verbose cli args?

File: lib/platform/bitbucket/README.md (hunk 3 above)
Comment (iamstarkov) [reply lib/platform/bitbucket/README.md]: why? it is internal BB adapter documentation

Comment (rarkins) [reply lib/platform/bitbucket/README.md]: now that you have implemented username/password, is there ever any need for users to manually encode anything?

Comment (iamstarkov) [reply lib/platform/bitbucket/README.md]: no, its not for renovates' users its for future BB adapter developers.

File: lib/config/definitions.js (hunk 2 above)
Comment (iamstarkov) [reply lib/config/definitions.js]: fixed

**Comment by iamstarkov** [comment]
fixed eslint, in turn it should fix the build

**Comment by iamstarkov** [comment]
i guess we need to add tests to `platform/index.js` and `config/index.js`

**Comment by rarkins** [comment]
Remaining:

- [ ] deleteBranch()
- [ ] tests/coverage

**Comment by rarkins** [comment]
Superseded by #2238

**Pull request closed without merging**